MailHelper class and /testconfig page

Vroo opened the pull request:
Take 2 on mail infra work, without random repeats of hint code.

Assigning to Kenny since he already started looking at this.

morganbr commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,117 @@
+                }
+                catch (FormatException)
+                {
+                    ;  // ignore
```
No semicolon required for this.

Vroo replied:
The `;` allows me to put a breakpoint there.

morganbr commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,117 @@
+    private void WriteMailToConsole(IEnumerable<string> recipients, string subject, string body)
+    {
+        // In dev environment, log the email instead of sending it if not configured.
+        Console.WriteLine("----------------------------------------");
```
I'm not sure if Console.WriteLine shows up anywhere. You might try Debug.WriteLine to have it show up in the debugger if it's not.

Vroo replied:
done

morganbr commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,117 @@
+        client.EnableSsl = true;
+        client.UseDefaultCredentials = false;
+        client.Credentials = new NetworkCredential(PublicSecret, PrivateSecret);
+        client.Send(msg);
```
It might be good to use SendMailAsync here and make MailHelper.Send an async method by changing the signature to ```async Task SendAsync(IEnumerable<string> recipients, bool bcc, string subject, string body)```

Vroo replied:
I haven't used the new async stuff in C# before. I understand the benefit of calling `SendMailAsync` instead. Since this function returns void, why would we need to change it to return a Task instead? Wouldn't the async call just work? What would we do with the Task?

Vroo replied:
Oh. I get it. If I do that then I don't need to wrap the call in `await (Task.Run())`?

Vroo replied:
It works fine if I switch to `SendMailAsync` and remove the `Task.Run`. I'll push it with that change. Let me know if it's wrong.

morganbr commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,117 @@
+            foreach (string recipient in recipients) {
+                try
+                {
+                    msg.To.Add(new MailAddress(recipient));
```
Should this either bcc players or send them separate messages in order to avoid reply-all problems?

Vroo replied:
Sending many emails will make us hit our email quota that much faster. Bcc is the way to avoid reply-all. Since the only case where we'd want them to see each other's emails is if one team member was sending mail to the rest of the team (which isn't contemplated). So I'll make bcc the only choice unless there's a single email.

tabascq replied:
I want a team application to *not* Bcc the team and the applicant. Can I do that?

morganbr commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,117 @@
+    
+    public static MailHelper Singleton { get; private set; }
+
+    public static void initialize(IConfiguration configuration, bool isDev)
```
Nit -- Capital first letter for methods in C# please.

Vroo replied:
oops.

morganbr commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,117 @@
+        Singleton = new MailHelper(configuration, isDev);
+    }
+
+    private MailHelper(IConfiguration configuration, bool isDev)
```
Looks like indentation might be off here or on the line below

Vroo replied:
done

morganbr commented:
Thanks, @Vroo ! I've put in a few minor comments and we'll be happy to merge after they're addressed.

morganbr commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,117 @@
+        if (bcc)
+        {
+            msg.To.Add(new MailAddress(ToAddressForBCC));
+            foreach (string recipient in recipients) {
```
Please put braces on the following line. If you're using visual studio, try running ctrl+k ctrl+f on this file, which will automatically format it to match our standards.

Vroo commented:
I just pushed an update. Please let me know if you have further comments. Thanks. @morganbr

tabascq commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            PublicSecret = configuration["Mailjet-Public"];
+            PrivateSecret = configuration["Mailjet-Private"];
+            Enabled = true;
```
I'm playing with adding mails for some scenarios and I found this line to be helpful:

            Enabled = PublicSecret != null && PrivateSecret != null;

Vroo replied:
Yeah, I could do it that way. I need the try/catch anyway, since `configuration[]` throws an exception if you haven't set the secrets and I don't know a way to say "get or default".

tabascq commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            Enabled = false;
+        }
+        Enabled = false;
```
That can't be right.

Vroo replied:
Fixed. I stuck that in to test something and forgot to take it out.

tabascq commented on ServerCore/Helpers/MailHelper.cs:
```diff
@@ -0,0 +1,128 @@
+        Debug.WriteLine("----------------------------------------");
+        foreach (string recipient in recipients)
+        {
+            Console.WriteLine("To: " + recipient);
```
Debug.WriteLine

Vroo replied:
done

Vroo commented:
For some reason, github won't let me reply to "I want a team application to not Bcc the team and the applicant. Can I do that?" so here it is:

A team application is when an unsolicited person requests to join a team? Since we don't have captains, you do want everyone to se it and for them all to see everyone else that got it. OK, that's a legit reason for non-bcc. I will add that capability back in with a new function.

tabascq commented:
I want to be able to do this:

            MailHelper.Singleton.Send(new string[] { Team.PrimaryContactEmail, LoggedInUser.Email },
                false,
                $"{LoggedInUser.Name} is applying to join {Team.Name}",
                $"To accept or reject this request, visit your team page.");

Vroo commented:
Again, can't reply directly. That's fine but it won't work the way the code is written now if they put multiple emails in the primary contact field. I'll go ahead and add that in here.

tabascq commented:
I have checked the entire PH20 registration list and there is never more than a single address. [Many teams have no primary address at all, but we should make this illegal.]

Vroo commented:
FYI, I had already implemented the comma/semicolon splitting before Kenny's comment and it was supported by the old server. I also had both html support and the unsubscribe footer in another branch and added them here, the latter since it's possible Kenny will get his code sending mail out deployed before mine.
